drivers/platform/aducm3029/flash.c

```diff
@@ -0,0 +1,289 @@
+			   ADI_FEE_MEMORY_SIZE,
+			   &adicup_extra->instance);
+	if(ret != ADI_FEE_SUCCESS)
+		goto error_adicup;
+
+	*device = dev;
+
+	return SUCCESS;
+
+error_adicup:
+	free(adicup_extra);
+error_dev:
+	free(dev);
+
+	return FAILURE;
+}
+
+/**
+ * Free memory allocated by flash_init().
+ *
+ * @param [in] dev - Pointer to the driver handler.
+ *
+ * @return 0 in case of success, error code otherwise.
+ */
+int32_t flash_remove(struct flash_dev *dev)
+{
+	int32_t ret;
+	struct adicup_flash_dev *adicup_extra;
+
+	if (!dev)
+		return FAILURE;
+
+	adicup_extra = dev->extra;
+
+	ret = adi_fee_Close(adicup_extra->instance);
+	if (ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	free(adicup_extra);
+	free(dev);
+
+	return SUCCESS;
+}
+
+/**
+ * Erase a flash page.
+ *
+ * @param [in] dev        - Pointer to the flash device handler.
+ * @param [in] page_no    - Page number.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t flash_clear_page(struct flash_dev *dev, int32_t page_no)
+{
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t fee_hw_error;
+	int32_t ret;
+
+	if (page_no > (dev->flash_size / dev->flash_page_size))
+		return FAILURE;
+
+	ret = adi_fee_PageErase(adicup_extra->instance, page_no, page_no,
+				&fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * Write a page in flash memory.
+ *
+ * @param [in] dev        - Pointer to the flash device handler.
+ * @param [in] page_no    - Page number.
+ * @param [in] data       - Pointer to the data to be written.
+ *                          data length must be FLASH_PAGE_SIZE_WORDS in size.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t flash_write_page(struct flash_dev *dev, int32_t page_no, uint32_t *data)
+{
+	ADI_FEE_TRANSACTION transaction;
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t fee_hw_error;
+	int32_t ret;
+
+	transaction.bUseDma = true;
+	transaction.nSize = FLASH_PAGE_SIZE_BYTES;
+	transaction.pWriteAddr = (uint32_t *)(page_no << FLASH_PAGE_ADDR_SHIFT);
+	transaction.pWriteData = data;
+
+	ret = adi_fee_Write(adicup_extra->instance, &transaction,
+			    &fee_hw_error);
+	if(ret != ADI_FEE_SUCCESS)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * Read-modify-write helper function for flash_write(). Since the smallest write
+ * unit for the ADuCM3029 is the page this function helps access data smaller than
+ * a page or accross multiple pages.
+ */
+static int32_t flash_read_then_write(struct flash_dev *dev, uint32_t flash_addr,
+				     uint32_t *array, uint32_t array_size)
+{
+	struct adicup_flash_dev *adicup_extra = dev->extra;
+	uint32_t page_nr, fee_hw_error;
+	uint32_t temp_ptr[FLASH_PAGE_SIZE_WORDS] __attribute__ ((aligned (sizeof(
+				uint32_t))));
+	int32_t ret;
+
+	if ((flash_addr & 0x3) != 0)
+		return FAILURE;
+	if (((FLASH_OFFSET_IN_PAGE(flash_addr) + array_size) * sizeof(uint32_t)) >
```

Comment:
if array is uin32_t  then array_size should be in words then` * sizeof(uint32_t) `shouldn't be needed.

Reply:
Or use uint8_t for array

Reply:
Same for the other functions.

Reply:
done